synapse/visibility.py

```diff
@@ -313,30 +322,39 @@ def _check_client_allowed_to_see_event(
     #
     # We can only do this check if the sender has *not* been erased, as if they
     # have we need to check the user's membership.
-    if not sender_erased and _check_history_visibility(event, visibility, is_peeking):
+    if (
+        not sender_erased
+        and _check_history_visibility(event, visibility, is_peeking)
+        == _CheckVisibility.ALLOWED
+    ):
         return event
 
-    allowed, membership = _check_membership(
-        user_id, event, visibility, state, is_peeking
-    )
-    if not allowed:
+    membership_result = _check_membership(user_id, event, visibility, state, is_peeking)
+    if not membership_result.allowed:
         return None
 
     # If the sender has been erased and the user was not joined at the time, we
     # must only return the redacted form.
-    if sender_erased and membership != Membership.JOIN:
+    if sender_erased and not membership_result.joined:
         event = prune_event(event)
 
     return event
 
 
+@attr.s(frozen=True, auto_attribs=True)
```

Comment:
I think we usually do `slots=True` (though I doubt it saves us much memory here)